packages/client/src/runtime/DataProxyEngine.ts

```diff
@@ -0,0 +1,282 @@
+  private async uploadSchema() {
+    const res = await fetch(this.url('schema'), {
+      method: 'PUT',
+      headers: this.headers,
+      body: this.schemaBase64,
+    })
+
+    if (res.ok) {
+      this.logEmitter.emit('info', {
+        message: `Schema (re)uploaded (hash: ${this.schemaHash})`,
+      })
+    } else {
+      this.logEmitter.emit('warn', { message: 'Could not upload the schema' })
+      throw new Error('Could not upload the schema')
+    }
+  }
+
+  request<T>(query: string, headers: Record<string, string>, attempt = 0) {
+    this.logEmitter.emit('query', { query })
+
+    return this.requestInternal<T>({ query, variables: {} }, headers, attempt)
+  }
+
+  async requestBatch<T>(
+    queries: string[],
+    headers: Record<string, string>,
+    isTransaction = false,
+    attempt = 0,
+  ) {
+    this.logEmitter.emit('query', {
+      query: `Batch${isTransaction ? ' in transaction' : ''} (${
+        queries.length
+      }):\n${queries.join('\n')}`,
+    })
+
+    const body = {
+      batch: queries.map((query) => ({ query, variables: {} })),
+      transaction: isTransaction,
+    }
+
+    const { batchResult } = await this.requestInternal<T>(
+      body,
+      headers,
+      attempt,
+    )
+
+    return batchResult
+  }
+
+  private async requestInternal<T>(
+    body: Record<string, any>,
+    headers: Record<string, string>,
+    attempt: number,
+  ) {
+    try {
+      this.logEmitter.emit('info', {
+        message: `Calling ${this.url('graphql')} (n=${attempt})`,
+      })
+
+      const res = await fetch(this.url('graphql'), {
+        method: 'POST',
+        headers: { ...headers, ...this.headers },
+        body: JSON.stringify(body),
+      })
+
+      // 404 on the GraphQL endpoint may mean that the schema
+      // was not uploaded yet.
+      if (res.status === 404) {
+        await this.uploadSchema()
+        throw new Error('Schema (re)uploaded')
+      }
+
+      if (!res.ok) {
+        throw new Error('GraphQL request failed')
+      }
+
+      return JSON.parse(await res.text())
+    } catch (err) {
+      if (attempt >= MAX_RETRIES) {
+        this.logEmitter.emit('error', {
+          message: `Failed to query: ${err.message}`,
+        })
+        throw err
+      } else {
+        const delay = await backOff(attempt)
+        this.logEmitter.emit('warn', { message: `Retrying after ${delay}ms` })
+        return this.requestInternal<T>(body, headers, attempt + 1)
+      }
+    }
+  }
+
+  // TODO: figure out how to support transactions
+  transaction(): Promise<any> {
+    throw new Error('Transactions are currently not supported in Data Proxy')
+  }
+}
+
+// Regex + prismafile feels fragile...
+// TODO: build a minimal single-purpose (URL extraction)
+// and well-tested TS schema parser
+function extractHostAndApiKey(schemaText: string) {
+  // Extract the datasource block so we feed prismafile with minimal input
+  const strippedToDatasource = /datasource[ \t]+[^\s]+[ \t]+\{[^}]+}/.exec(
```

Comment:
This feels slightly unnecessary when we parse it a few steps later. I guess it's an optimization, but probably not a very good one since it scans the whole string again anyway.